src/wui/info_panel.h

```diff
@@ -0,0 +1,129 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+#ifndef WL_WUI_INFO_PANEL_H
+#define WL_WUI_INFO_PANEL_H
+
+#include <list>
+#include <memory>
+
+#include "logic/message_id.h"
+#include "ui_basic/dropdown.h"
+#include "ui_basic/textarea.h"
+#include "wui/logmessage.h"
+
+class InfoPanel;
+class InteractiveBase;
+class InteractivePlayer;
+class MainToolbar;
+
+namespace Widelands {
+struct Message;
+struct MessageQueue;
+}  // namespace Widelands
+
+class MessagePreview : public UI::Textarea {
+public:
+	MessagePreview(InfoPanel&, const Widelands::Message&, Widelands::MessageId);
+	MessagePreview(InfoPanel&, const std::string& text, const std::string& tooltip);
+	~MessagePreview() override {
+	}
+
+	void think() override;
+	void draw(RenderTarget&) override;
+	bool handle_mousepress(uint8_t, int32_t, int32_t) override;
+
+private:
+	InfoPanel& owner_;
+	uint32_t creation_time_;
+
+	bool message_still_exists() const;
+
+	const Widelands::Message* message_;
+	Widelands::MessageId id_;
+};
+
+class InfoPanel : public UI::Panel {
+public:
+	InfoPanel(InteractiveBase&);
+	~InfoPanel() override {
+	}
+
+	// Update the text area without relayouting
+	void set_time_string(const std::string&);
+	void set_speed_string(const std::string&);
+	void set_fps_string(const std::string&);
+	void set_coords_string(const std::string&);
+
+	void log_message(const std::string&);
+
+	void think() override;
+	void layout() override;
+
+	void set_toolbar(MainToolbar&);
+
+	void fast_forward_message_queue();
+
+	void draw(RenderTarget&) override;
+	bool handle_mousepress(uint8_t, int32_t, int32_t) override;
+	bool handle_mouserelease(uint8_t, int32_t, int32_t) override;
+	bool handle_mousemove(uint8_t, int32_t, int32_t, int32_t, int32_t) override;
+
+private:
+	friend class MessagePreview;
+
+	InteractiveBase& ibase_;
+	InteractivePlayer* iplayer_;
+
+	bool on_top_;
+
+	enum DisplayMode {
+		kCmdSwap = 1,
+		kPinned = 2,
+		kMinimized = 4,
+		kOnMouse_Visible = 8,
+		kOnMouse_Hidden = 16
+	};
+	DisplayMode display_mode_;
+	void update_mode();
+	void rebuild_dropdown();
+
+	Vector2i last_mouse_pos_;
+	bool is_mouse_over_panel() const;
+	void set_textareas_visibility(bool);
+
+	void update_time_speed_string();
+
+	MainToolbar* toolbar_;
+
+	UI::Dropdown<DisplayMode> toggle_mode_;
+	UI::Textarea text_time_speed_, text_fps_, text_coords_;
+
+	std::list<MessagePreview*> messages_;
+	size_t index_of(const MessagePreview&) const;
+	void pop_message(MessagePreview&);
+	void push_message(MessagePreview&);
+	std::unique_ptr<Notifications::Subscriber<LogMessage>> log_message_subscriber_;
+	const Widelands::MessageQueue* message_queue_;
+	std::unique_ptr<Widelands::MessageId> last_message_id_;
```

Comment:
MessageId is not a large type (just an `uint32_t`), so using a literal here shouldn't be a problem.
```suggestion
	Widelands::MessageId last_message_id_;
```

Reply:
In line 432/433 I need a special value indicating that `last_message_id_` has not been initialized yet, and this value needs to be different from `MessageId::null()`. Since we're on C++11 with no `std::optional` available a unique_ptr is the best way to model this.